### ISSUE TITLE
Modifying the Id of a persistent entity to use serial4 instead of serial8

### ISSUE THREAD

**Issue by codygman** [yesodweb/persistent#1336]
I have this:

```
fact
    Id Int32 sqltype=serial4
```

And that creates this:

```
CREATE TABLE public.fact (
    id integer NOT NULL
)
```

but not the sequence:

```
ALTER TABLE ONLY public.viewing_fact ALTER COLUMN id SET DEFAULT nextval('public.viewing_fact_id_seq'::regclass);
```

I guess only serial8 is supported in persistent after finding [maySerial](https://github.com/yesodweb/persistent/blob/fb7a42f3f3f8cecaa45b6e32cda37d495596a908/persistent-postgresql/Database/Persist/Postgresql.hs#L677)?

I think this is related to [this comment of yours](https://github.com/yesodweb/persistent/issues/1037#issuecomment-594211639) and that idea is blocked by #1321 maybe?

Info:
```
/bin/uname
Linux 01cd201e6c60 5.11.21 #1-NixOS SMP Fri May 14 08:50:20 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
/usr/local/bin/stack
Version 2.7.3, Git revision 7927a3aec32e2b2e5e4fb5be76d0d50eddcc197f x86_64 hpack-0.34.4
/usr/local/bin/stack
The Glorious Glasgow Haskell Compilation System, version 8.10.7
/usr/local/bin/stack
```

**Comment by parsonsmatt** [comment]
Hmm. I'd expect the `sqltype=` to override whatever is provided here. If `serial4` were properly given as the type in the `CREATE TABLE` statement, then this should work just fine.

**Comment by codygman** [comment]
The create table statement doesn't contain serial4. So given:

```
fact
    Id Int32 sqltype=serial4
```

It generates:

```
Migrating: CREATe TABLE "fact"("id" INT4 PRIMARY KEY UNIQUE)
```

But I would expect:

```
Migrating: CREATe TABLE "fact"("id" SERIAL4 PRIMARY KEY UNIQUE)
```

**Comment by parsonsmatt** [comment]
Yeah, precisely. The bug appears to be in how we're tracing the `sqltype` and it's getting overridden somewhere.

**Comment by codygman** [comment]
Going to workaround with this for now:

```
migrations :: Persist.Migration
migrations = do
  -- This is because for some reason persistent doesn't seem to respect a custom sqltype for id types
  -- even though it should: https://github.com/yesodweb/persistent/issues/1336
  Persist.addMigration False $ mconcat
    [ "CREATE SEQUENCE if not exists public.fact_id_seq "
    , "AS integer "
    , "START WITH 1 "
    , "INCREMENT BY 1 "
    , "NO MINVALUE "
    , "NO MAXVALUE "
    , "CACHE 1; "
    , "ALTER TABLE ONLY public.fact ALTER COLUMN id SET DEFAULT nextval('public.fact_id_seq'::regclass)"
    ]
```

**Comment by codygman** [comment]
> Yeah, precisely. The bug appears to be in how we're tracing the `sqltype` and it's getting overridden somewhere.

Something interesting to note is that the fact_id_seq gets created but is just never used.

**Comment by codygman** [comment]
> Going to workaround with this for now:
> 
> ```haskell
> {-
> -- fact.persistentmodels
> fact
>     Id Int32
> -}
> 
> migrations :: Persist.Migration
> migrations = do
>   -- This is because for some reason persistent doesn't seem to respect a custom sqltype for id types
>   -- even though it should: https://github.com/yesodweb/persistent/issues/1336
>   Persist.addMigration False $ mconcat
>     [ "CREATE SEQUENCE if not exists public.fact_id_seq "
>     , "AS integer "
>     , "START WITH 1 "
>     , "INCREMENT BY 1 "
>     , "NO MINVALUE "
>     , "NO MAXVALUE "
>     , "CACHE 1; "
>     , "ALTER TABLE ONLY public.fact ALTER COLUMN id SET DEFAULT nextval('public.fact_id_seq'::regclass)"
>     ]
> ```

I'm a little concerned with this workaround because in a test I have I'm getting an error of:


```
    • No instance for (Database.Persist.Class.PersistStore.ToBackendKey
                         Database.Persist.SqlBackend.Internal.SqlBackend
                         Factt)
```

So as defined the `ToBackendKey` isn't generated for this type by persistent I guess? I wanted to use this workaround to unblock our production deploys but this makes me hesitant.

**Comment by parsonsmatt** [comment]
Yeah - `ToBackendKey` is a class that gives you, essentially, a pair of functions that map from a `Key a` to `Int64`, via `toBackendKey :: Key a -> BackendKey SqlBackend`, and `Backendkey SqlBackend = SqlBackendKey { unSqlBackendKey :: Int64 }`. So if you're not properly an `Int64` then this won't work.

The fix would be to write a specific key constructor instead of `fromSqlKey` or `toSqlKey`

**Comment by codygman** [comment]
Alright, it's fine for me to remove that then since I only really use `toBackendKey` to test backwards compatibility between Persistent and the library we are migrating from.
